File: validateBST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """

        # Time Complexity: O(N)
        # Space Complexity: O(N)
        def helper(root, left, right):
            if not root: return True
            if not (left < root.val < right): return False
            return helper(root.left, left, root.val) and helper(root.right, root.val, right)

        return helper(root, -float("inf"), float("inf"))

File: ValidateBST.cs
// Time Complexity : O(n)
// Space Complexity : O(h)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

//min and max approach, we store min and max at ever node
//compare with root.val
public bool IsValidBST(TreeNode root) {
    if(root == null)
        return true;
    
    return helper(root, null, null);
}

private bool helper(TreeNode root, int? min, int? max)
{
    //base condition
    if(root == null)
        return true;
    
    if(min >= root.val)
        return false;
    
    if(max <= root.val)
        return false;
    
    var case1 = helper(root.left, min, root.val);
    var case2 = helper(root.right, root.val, max);

    return case1 && case2;
}


//iterative approach using stack
//we compare the root value with left noe value, if its smaller, return false
//O(n), O(n)
public bool IsValidBST(TreeNode root) {
    if(root == null)
        return true;

    Stack<TreeNode> stack = new Stack<TreeNode>();
    TreeNode prev = null;
    while(root != null || stack.count >0)
    {
        while(root != null)
        {
            stack.Push(root);
            root = root.left;
        }
        root = stack.Pop();
        if(prev != null && prev.val >= root.val)
            return false;
        prev = root;
        root = root.right;
    }
}


//recurson approach
//O(n)
//O(h) h is hegiht of tree
TreeNode prev = null;
bool isValid;
public bool IsValidBST(TreeNode root) {
    if(root == null)
        return true;

    isValid = false;
    inOrder(root);
    return isValid;
}

private bool inOrder(TreeNode root)
{
    //base condition
    if(root == null)
        return true;
    
    //logic
    if(inOrder(root.left) == false)
        return false;
    
    if(prev != null && prev.val >= root.val)
        return false;

    prev = root;
    return inOrder(root.right);

}